File: 84. Largest Rectangle in Histogram.py
# https://leetcode.com/problems/largest-rectangle-in-histogram/

# Given an array of integers heights representing the histogram's bar height where
# the width of each bar is 1, return the area of the largest rectangle in the
# histogram.

################################################################################

# for each ele, find the frist ele on left and right that is smaller
# use stack to find numbers faster

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        if len(heights) == 1: return heights[0]
                
        heights.append(0) # for popping all prev ele
        ans = 0
        stack = [0] # store idx
        
        for i in range(1, len(heights)):
            while stack and heights[stack[-1]] > heights[i]: # pop the prev bar since found its right limit
                idx = stack.pop() # calculate area of this bar
                height = heights[idx]
                
                if stack:
                    prev_idx = stack[-1]
                else:
                    prev_idx = - 1
                    
                width = i - prev_idx - 1 # left limit idx = prev_idx, right limit idx = i
                ans = max(ans, height * width)
            stack.append(i)
        return ans

File: 84. Largest Rectangle in Histogram.cpp
// https://leetcode.com/problems/largest-rectangle-in-histogram/

// Given an array of integers heights representing the histogram's bar height where
// the width of each bar is 1, return the area of the largest rectangle in the
// histogram.

////////////////////////////////////////////////////////////////////////////////

// find largest rectangle at each index i
// need index of prev and next less num -> monostack
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int n = heights.size();
        vector<int> prevLessIdx(n, -1), nextLessIdx(n, n);
        stack<int> toRight, toLeft;
        for (int i = 0; i < n; ++i) {
            // pop if not less
            while (!toRight.empty() and heights[toRight.top()] >= heights[i]) {
                toRight.pop();
            }
            prevLessIdx[i] = toRight.empty() ? -1 : toRight.top();
            toRight.push(i);
        }
        for (int i = n - 1; i >= 0; --i) {
            // pop if not less
            while (!toLeft.empty() and heights[toLeft.top()] >= heights[i]) {
                toLeft.pop();
            }
            nextLessIdx[i] = toLeft.empty() ? n : toLeft.top();
            toLeft.push(i);
        }
        
        int maxArea = 0;
        for (int i = 0; i < n; ++i) {
            maxArea = max(maxArea, heights[i] * (nextLessIdx[i] - prevLessIdx[i] - 1));
        }
        return maxArea;
    }
};
